mitogen/master.py

```diff
@@ -706,21 +706,20 @@ def on_shutdown(self, broker):
     def _first_stage():
         import os,sys,zlib
         R,W=os.pipe()
-        R2,W2=os.pipe()
+        r,w=os.pipe()
         if os.fork():
             os.dup2(0,100)
             os.dup2(R,0)
-            os.dup2(R2,101)
-            for f in R,R2,W,W2: os.close(f)
-            os.environ['ARGV0'] = sys.executable
-            os.execv(sys.executable,['mitogen:CONTEXT_NAME'])
-        else:
-            os.write(1, 'EC0\n')
-            C = zlib.decompress(sys.stdin.read(input()))
-            os.fdopen(W,'w',0).write(C)
-            os.fdopen(W2,'w',0).write('%s\n%s' % (len(C),C))
-            os.write(1, 'EC1\n')
-            sys.exit(0)
+            os.dup2(r,101)
+            for f in R,r,W,w:os.close(f)
+            os.environ['ARGV0']=e=sys.executable
+            os.execv(e,['mitogen:CONTEXT_NAME'])
+        os.write(1,'EC0\n')
+        C=zlib.decompress(sys.stdin.read(input()))
```

Comment:
Is this line correct for Python 3.x?

AIUI `input()` evaluates the first line as an integer, which is passed to `sys.stdin.read()` as the number of bytes to request. However on Python 3 `input()` does what `raw_input()` does on Python 2. So on Python 3 this line is passing a string as the length argument to `sys.stdin.read()`.

Reply:
Also, could the use of `sys.stdin.read(...)` be replaced with `os.read(1, ...)`?